packages/ai-workspace-agent/src/browser/functions.ts

```diff
@@ -39,15 +53,59 @@ export class WorkspaceFunctionScope {
             throw new Error('Access outside of the workspace is not allowed');
         }
     }
-    /**
-     * Determines whether a given file or directory should be excluded from workspace operations.
-     *
-     * @param stat - The `FileStat` object representing the file or directory to check.
-     * @returns `true` if the file or directory should be excluded, `false` otherwise.
-     */
-    shouldExclude(stat: FileStat): boolean {
-        const excludedFolders = ['node_modules', 'lib'];
-        return stat.resource.path.base.startsWith('.') || excludedFolders.includes(stat.resource.path.base);
+
+    private async initializeGitignoreWatcher(workspaceRoot: URI): Promise<void> {
+        if (this.gitignoreWatcherInitialized) {
+            return;
+        }
+
+        const gitignoreUri = workspaceRoot.resolve(this.GITIGNORE_FILE_NAME);
+        this.fileService.watch(gitignoreUri);
+
+        this.fileService.onDidFilesChange(async event => {
+            if (event.contains(gitignoreUri)) {
+                this.gitignoreMatcher = undefined;
+            }
+        });
+
+        this.gitignoreWatcherInitialized = true;
+    }
+
+    async shouldExclude(stat: FileStat): Promise<boolean> {
+        const considerGitIgnore = this.preferences.get(CONSIDER_GITIGNORE_PREF, false);
+        const userExcludes = this.preferences.get<string[]>(USER_EXCLUDES_PREF, []);
+
+        const fileName = stat.resource.path.base;
+
+        for (const pattern of userExcludes) {
+            const matcher = new Minimatch(pattern, { dot: true });
+            if (matcher.match(fileName)) { return true; }
+        }
+
+        if (considerGitIgnore) {
+            const workspaceRoot = await this.getWorkspaceRoot();
+            await this.initializeGitignoreWatcher(workspaceRoot);
+
+            const gitignoreUri = workspaceRoot.resolve(this.GITIGNORE_FILE_NAME);
+
+            try {
+                const fileStat = await this.fileService.resolve(gitignoreUri);
+                if (fileStat && !fileStat.isDirectory) {
+                    if (!this.gitignoreMatcher) {
+                        const gitignoreContent = await this.fileService.read(gitignoreUri);
+                        this.gitignoreMatcher = ignore().add(gitignoreContent.value);
+                    }
+                    const relativePath = workspaceRoot.relative(stat.resource);
+                    if (relativePath && this.gitignoreMatcher.ignores(relativePath.toString())) {
+                        return true;
+                    }
+                }
+            } catch (error) {
```

Comment:
```suggestion
            } catch {
```

We don't need the error as a variable in that case.